diagnostics/buffer.go

```diff
@@ -0,0 +1,124 @@
+package diagnostics
+
+import (
+	"sync/atomic"
+
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/topdown"
+)
+
+var seq uint64
+
+// Buffer stores diagnostic information. They are not safe for concurrent access.
+type Buffer struct {
```

Comment:
Due to offline discussion, leaving the buffer implementation. Will add boilerplate copyright, but moving into the server so the package comment is not needed.